File: DSA/Java/Complete DFS Traversal for Binary Tree (Recursive).java
// Here in this program we will perform Preorder, Inorder, Postorder Traversal all at once.

import java.util.Scanner;

class Node {
    int val;
    Node left, right;

    Node(int data) {
        val = data;
        left = null;
        right = null;
    }
}

public class BinaryTreeTraversals {
    static Node buildTree() {
        Scanner scanner = new Scanner(System.in);
        System.out.print("Enter data (-1 to stop): ");
        int data = scanner.nextInt();

        if (data == -1) {
            return null;
        }

        Node root = new Node(data);

        System.out.println("Enter data in left of " + data);
        root.left = buildTree();

        System.out.println("Enter data in right of " + data);
        root.right = buildTree();

        return root;
    }

    static void preorder(Node root) {
        if (root == null) {
            return;
        }
        System.out.print(root.val + " ");
        preorder(root.left);
        preorder(root.right);
    }

    static void inorder(Node root) {
        if (root == null) {
            return;
        }
        inorder(root.left);
        System.out.print(root.val + " ");
        inorder(root.right);
    }

    static void postorder(Node root) {
        if (root == null) {
            return;
        }
        postorder(root.left);
        postorder(root.right);
        System.out.print(root.val + " ");
    }

    public static void main(String[] args) {
        Node root = buildTree();

        System.out.println("Preorder Traversal is:");
        preorder(root);
        System.out.println();

        System.out.println("Inorder Traversal is:");
        inorder(root);
        System.out.println();

        System.out.println("Postorder Traversal is:");
        postorder(root);
        System.out.println();
    }
}

File: DSA - Data Structures And Algorithm/C++/Binary Tree/Complete DFS Traversal for Binary Tree (Recursive).cpp
// Here in this program we will perform Preorder, Inorder, Postorder Traversal all at once.

#include <iostream>
using namespace std;

class Node{
  public:
  int val;
  Node* left;
  Node* right;
  Node (int data){
    val = data;
    left = NULL;
    right = NULL;
  }
};

Node* BuildTree(Node* &root){
  int data;
  cout<<"Enter data:- ";
  cin>>data;
  root = new Node(data);
  if (data == -1) return NULL;

  cout<<"Enter data in left of "<<data<<endl;
  root->left = BuildTree(root->left);
  
  cout<<"Enter data in right of "<<data<<endl;
  root->right = BuildTree(root->right);

  return root;
}

void Preorder(Node* &root){
  if (root==NULL) return;
  cout<<root->val<<" ";
  Preorder(root->left);
  Preorder(root->right);
}

void Inorder(Node* &root){
  if (root==NULL) return;
  Inorder(root->left);
  cout<<root->val<<" ";
  Inorder(root->right);
}

void Postorder(Node* &root){
  if (root==NULL) return;
  Postorder(root->left);
  Postorder(root->right);
  cout<<root->val<<" ";
}

int main(){
  Node* root = NULL;
  BuildTree(root);
  
  cout<<"Preorder Traversal is:- ";
  Preorder(root);
  cout<<endl;
  
  cout<<"Inorder Traversal is:- ";
  Inorder(root);
  cout<<endl;
  
  cout<<"Postorder Traversal is:- ";
  Postorder(root);
  cout<<endl;
}
